shared/chat/conversation/giphy/width.js

```diff
@@ -1,10 +1,221 @@
 // @flow
 
+const minWidth = 100
+const maxWidth = 200
+
 export const scaledWidth = (width: number) => {
   return width * 0.5
 }
 
-export const getMargin = (width: number, maxWidth: number) => {
-  const m = -((scaledWidth(width) - maxWidth) / 2)
-  return m > 0 ? 0 : m
+// getMargins implements a simple algorithm to resize GIF previews given a dynamic display size, and
+// dynamic sized width GIFs. The basic idea is to try to form rows by either compressing down GIFs to fit,
+// or expanding to fit, and see which method works better.
+export const getMargins = (totalWidth: number, widths: Array<number>) => {
+  const images = []
+  widths.forEach(w => {
+    images.push(new Image(scaledWidth(w), 0))
+  })
+  // set to the desired size, some GIFs are very wide and take up too much space initially. We will allow
+  // the expansion phase to go over this clamp however.
+  clampAtMaxWidths(images)
+  let res = []
+  let longRow = []
+  for (let index = 0; index < images.length; index++) {
+    const im = images[index]
```

Comment:
since `index` is reassigned from `pickRow` probably worth a sanity check the index is within bounds here